### ISSUE TITLE
`@stylistic/unit-case` reports at the word rather than at the unit

### ISSUE THREAD

**Issue by firefoxic** [stylelint-stylistic/stylelint-stylistic#296]
Found while reviewing the branch that closes [#234](https://github.com/stylelint-stylistic/stylelint-stylistic/issues/234). Every shape below reproduces on `main` before that branch as it does after it: the branch narrows the warning where a bang flag stands, and leaves these three alone.

### What minimal example or steps are needed to reproduce the bug?

Three shapes, each reported at a position that is not the unit's:

```css
a { b: 10px*2REM; }
a { b: 10px\0; }
a { b: calc(100% - #{$m * 2px}); }
```

### What minimal configuration is needed to reproduce the bug?

```json
{
  "plugins": [
    "@stylistic/stylelint-plugin"
  ],
  "rules": {
    "@stylistic/unit-case": "lower"
  }
}
```

The third wants `postcss-scss` and the option `"upper"`.

### How did you run Stylelint?

```shell
stylelint "**/*.css" --config myconfig.json
```

### What did you expect to happen?

Each warning to stand at the unit it names, and to end where that unit ends.

### What actually happened?

```
a { b: 10px*2REM; }                  lower
  1:19-1:20   underlines `}`, past the end of the value
  1:10-1:17   underlines `px*2REM`

a { b: 10px\0; }                     upper
  1:10-1:14   underlines `px\0`, though the message names `px`

a { b: calc(100% - #{$m * 2px}); }   upper, postcss-scss
  1:28-1:31   underlines `px}`, though the message names `px`
```

The first is not merely misplaced: its `endIndex` is smaller than its `index`, so the range runs backwards.

### What is the cause?

Two readings, both in `lib/rules/unit-case/index.js`.

The warning ends at `valueNode.sourceEndIndex`, the end of the **word**, while the unit `getDimension` hands back is read out of a copy with the interpolation blurred out and the `\0` and `\9` hacks stripped. Whatever those two took off the end of the word is left inside the run the warning underlines.

The parts of a multiplication are worse off. The `value.includes("*")` branch hands `processValue` a node it builds itself, with a `sourceIndex` of `value.indexOf(val) + val.length + 1` — a position that is not where the part stands, and that for the last part of the word exceeds the word's own end — while `sourceEndIndex` is spread over from the whole word. So the two positions of such a node describe no text at all.

### Do you have a proposal to fix the bug?

Count both positions of a warning from the unit inside the node rather than from the node's end, and give the parts of a multiplication positions that are their own. The second half is entangled with [#297](https://github.com/stylelint-stylistic/stylelint-stylistic/issues/297), which is about what that branch reads rather than about where it reports.
